[PyTorch] Adding example for Iterative pruning (#3884)
mlflow/mlflow: examples/pytorch/IterativePruning/README.md

## Iterative Pruning
Pruning is the process of compressing a neural network that involves removing weights from a trained model.

> It could be achieved by either removing the neurons in a specific layer or make the weights of the connections 

Reviewer: ```suggestion
Example pruning techniques include removing the neurons within a specific layer, or setting the weights of connectons that are already near zero to zero. This script applies the latter technique. Pruning a model reduces its size, at the cost of worsened model accuracy.
```

Proofreading suggestion - also, is there a definition of iterative pruning from the pytorch docs we could link to here?

Reviewer 2: Updated the string

Reviewer 2: Added a link to https://pytorch.org/tutorials/intermediate/pruning_tutorial.html

that are nearer to zero as zero. The script is an example of later. Pruning a model has an impact on
accuracy of the model, while it makes the model lightweight.

> In this example, we train a model to classify MNIST handwritten Digit recognition dataset. Basemodel along with the parameters, metrics and summary are stored in mlflow.

Reviewer: ```suggestion
In this example, we train a model to classify MNIST handwritten digit recognition dataset, and then apply iterative pruning to compress the model. The initial model ("base model") along with its parameters, metrics and summary are stored in mlflow.
```

Reviewer 2: Updated the string

> Subsequently, the trained MNIST model stored in mlflow is saved and pruned iteratively by using the custom 

Reviewer: ```suggestion
Subsequently, the base model is pruned iteratively by using the custom 
```

Reviewer 2: Updated the string

> inputs provided from the cli. Ax is a platform for optimizing any kind of experiment, including machine learning experiments,

Reviewer: How and why does Ax come into play? I'm kind of confused - I thought this example was about iterative pruning?

Reviewer: Ah, I see we mention that we use Ax for pruning later on. Could we state upfront here that we leverage Ax for iterative pruning to avoid ^such confusion?

A/B tests, and simulations. [Ax](https://ax.dev/docs/why-ax.html) can optimize discrete configurations using multi-armed bandit optimization,
and continuous (e.g., integer or floating point)-valued configurations using Bayesian optimization.

> The objective function of the experiment trails is "test_accuracy" based on which the model is evaluated at each trial and the best set of parameters are derived.

Reviewer: ```suggestion
The objective function of the experiment trials is "test_accuracy" based on which the model is evaluated at each trial and the best set of parameters are derived.
```

Reviewer 2: fixed typo

AXClient is used to provide the initial pruning percentage as well as decides the number
of trails to be run. The summary of the pruned model is captured in a separate file and stored as an artifact in MLflow.

> ### Setting up Environment variables.

Reviewer: Most users shouldn't need to do this right? Could we move `Setting up Environment variables` to a subsection after `Running the code to Iteratively Prune the Trained Model`? We should also reword the header to make it clearer to users why they should "Set up environment variables". Maybe we could use a header like "Log runs to a custom experiment" instead?

One way to think of it is that setting the environment variables is the specific implementation we're recommending for logging runs to a custom experiment

Reviewer 2: I have removed this section. The initial logic had the iterative pruning process in two phases. There were two different scripts. One for the base model which trains and stored into mlflow and other script which performs the iterative pruning process. The second script fetches the base model based on the experiment name. I found it really confusing and also its error prune - what if there are multiple base runs in the experiment.

Hence, i have updated the example run in a single flow. We no longer need this section.
